guidance/library/_pydantic.py

```diff
@@ -0,0 +1,53 @@
+import inspect
+from typing import Any, Dict, Optional, Type, Union
+
+from pydantic import BaseModel, TypeAdapter
+from pydantic.json_schema import GenerateJsonSchema
+
+import guidance
+
+from ._json import json as gen_json
+
+
+class GenerateJsonSchemaSafe(GenerateJsonSchema):
+    """
+    Subclass pydantic's GenerateJsonSchema to catch pydantic schemas that will not
+    translate properly to json schemas used for generation.
+
+    In particular, JSON schemas do not offer a way to specify "key type",
+    so we need to raise an exception if users attempt to specify non-string
+    keys through pydantic. Otherwise, they may get unexpected output from
+    model generation.
+    """
+
+    def generate_inner(self, schema):
+        if schema["type"] == "dict":
+            key_type = schema["keys_schema"]["type"]
+            if key_type != "str":
+                raise TypeError(
+                    f"JSON does not support non-string keys, got type {key_type}"
+                )
+        return super().generate_inner(schema)
```

Comment:
What type of non-string keys does Pydantic allow for, out of curiosity? Not necessary for this PR, but curious if we should perhaps in the future explore casting to a string as an intermediary, and then replacing with the original type after we finish generating? Seems messy but might expand coverage of the JSON types we can handle -- particularly if these keys aren't model generated. Curious about your thoughts here, but again not critical for this PR

Reply:
Great question.

Pydantic is a lot more powerful than JSON in terms of what can be validated. AFAIK, there's little to no restriction on what types you can validate (so long as they are builtins and unions/subscripted versions thereof). 

A quick check just demonstrated to me that it doesn't even check that you use hashable keys for dicts, e.g. you can specify `dict[dict, int]` just fine; you just won't be able to construct an object that will validate correctly.

It will also do something very similar to what you suggested (converting to strings) if you try serializing objects that can't typically be serialized with vanilla JSON. The primary problem that `guidance` faces if it wants to replicate this behavior is that the fact that you wanted some type other than strings as your keys just _cannot be expressed_ in the language of JSON schemas:

![non_string_keys](https://github.com/guidance-ai/guidance/assets/927410/2b8eb162-1706-4315-90b7-034603451b54)

Because we're using JSON schemas as the language for describing the grammar, we're stuck with this lossy schema that can't encode the constraints that the user desired. I figured it was safer to throw an exception than to pass silently and end up with arbitrary string keys that don't conform to user expectations.

The alternative would be to either use something other than JSON schemas as the "backend" to generation in this context or to extend the JSON schema specification to something more powerful. You can see that pydantic actually does something like that with "core schemas":

![non_string_keys_core_schema](https://github.com/guidance-ai/guidance/assets/927410/880c4f97-7d5e-44e1-8ab3-d01050147444)

Finally, it seems to me that pydantic has some trouble with "round trip" loading of an object dumped like this. They accept a `round_trip` kwd to `dump_json`, but it doesn't seem to be working the way I would expect. Definitely interested in taking a deeper look at some point. 

Anyway, for the sake of this PR, I definitely recommend raising the exception in this case.

End of rant :laughing:

Reply:
Addendum to the rant: it may be reasonable to add support for generating "python literals", i.e. strings that can be evaluated by `ast.literal_eval`, maybe using pydantic core schemas as the specification backend.

The primary use case I'm imagining is to generare arguments for Python functions with type-annotated arguments. I may write a mini RFC or a toy implementation and do an interest-check at some point in the future :)